src/pkg/oci/pull.go

```diff
@@ -0,0 +1,241 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2021-Present The Zarf Authors
+
+// Package oci contains functions for interacting with Zarf packages stored in OCI registries.
+package oci
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"sync"
+
+	"github.com/defenseunicorns/zarf/src/config"
+	"github.com/defenseunicorns/zarf/src/pkg/message"
+	"github.com/defenseunicorns/zarf/src/pkg/utils"
+	"github.com/defenseunicorns/zarf/src/types"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/content/file"
+)
+
+var (
+	// AlwaysPull is a list of paths that will always be pulled from the remote repository.
+	AlwaysPull = []string{config.ZarfYAML, config.ZarfChecksumsTxt, config.ZarfYAMLSignature}
+)
+
+// LayersFromPaths returns the descriptors for the given paths from the root manifest.
+func (o *OrasRemote) LayersFromPaths(requestedPaths []string) (layers []ocispec.Descriptor, err error) {
+	manifest, err := o.FetchRoot()
+	if err != nil {
+		return nil, err
+	}
+	for _, path := range requestedPaths {
+		layer := manifest.Locate(path)
+		if o.isEmptyDescriptor(layer) {
+			return nil, fmt.Errorf("path %s does not exist in this package", path)
+		}
+		layers = append(layers, layer)
+	}
+	return layers, nil
+}
+
+// LayersFromRequestedComponents returns the descriptors for the given components from the root manifest.
+//
+// It also retrieves the descriptors for all image layers that are required by the components.
+//
+// It also respects the `required` flag on components, and will retrieve all necessary layers for required components.
+func (o *OrasRemote) LayersFromRequestedComponents(requestedComponents []string) (layers []ocispec.Descriptor, err error) {
+	root, err := o.FetchRoot()
+	if err != nil {
+		return nil, err
+	}
+
+	pkg, err := o.FetchZarfYAML(root)
+	if err != nil {
+		return nil, err
+	}
+	images := map[string]bool{}
+	tarballFormat := "%s.tar"
+	for _, name := range requestedComponents {
+		component := utils.Find(pkg.Components, func(component types.ZarfComponent) bool {
+			return component.Name == name
+		})
+		if component.Name == "" {
+			return nil, fmt.Errorf("component %s does not exist in this package", name)
+		}
+	}
+	for _, component := range pkg.Components {
+		// If we requested this component, or it is required, we need to pull its images and tarball
+		if utils.SliceContains(requestedComponents, component.Name) || component.Required {
+			for _, image := range component.Images {
+				images[image] = true
+			}
+			layers = append(layers, root.Locate(filepath.Join(config.ZarfComponentsDir, fmt.Sprintf(tarballFormat, component.Name))))
+		}
+	}
+	if len(images) > 0 {
+		// Add the image index and the oci-layout layers
+		layers = append(layers, root.Locate(root.indexPath), root.Locate(root.ociLayoutPath))
+		// Append the sbom.tar layer if it exists
+		sbomDescriptor := root.Locate(config.ZarfSBOMTar)
```

Comment:
components can also contain SBOMs (not just images).

Reply:
eh? idk what you mean here.

Reply:
ah you mean a component could have `file` SBOMs but no image SBOMs